server/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryIT.java

```diff
@@ -161,78 +148,46 @@ public void testNodeFailuresAreProcessedOnce() throws ExecutionException, Interr
-        final AtomicReference<Exception> reference = new AtomicReference<>();
-        internalCluster().getInstance(TransportService.class, noneMasterNode).sendRequest(node, PublishClusterStateAction.SEND_ACTION_NAME,
-                new BytesTransportRequest(bytes, Version.CURRENT), new EmptyTransportResponseHandler(ThreadPool.Names.SAME) {
-
-            @Override
-            public void handleResponse(TransportResponse.Empty response) {
-                super.handleResponse(response);
-                latch.countDown();
-            }
-
-            @Override
-            public void handleException(TransportException exp) {
-                super.handleException(exp);
-                reference.set(exp);
-                latch.countDown();
-            }
-        });
-        latch.await();
-        assertThat(reference.get(), notNullValue());
-        assertThat(ExceptionsHelper.detailedMessage(reference.get()),
-                containsString("cluster state from a different master than the current one, rejecting"));
-    }
-
-    public void testHandleNodeJoin_incompatibleClusterState() throws UnknownHostException {
-        String masterOnlyNode = internalCluster().startMasterOnlyNode();
+    public void testHandleNodeJoin_incompatibleClusterState()
+            throws IOException, InterruptedException, ExecutionException, TimeoutException {
+        internalCluster().startMasterOnlyNode();
         String node1 = internalCluster().startNode();
-        ZenDiscovery zenDiscovery = (ZenDiscovery) internalCluster().getInstance(Discovery.class, masterOnlyNode);
         ClusterService clusterService = internalCluster().getInstance(ClusterService.class, node1);
+        TransportService transportService = internalCluster().getInstance(TransportService.class, node1);
         final ClusterState state = clusterService.state();
         MetaData.Builder mdBuilder = MetaData.builder(state.metaData());
         mdBuilder.putCustom(CustomMetaData.TYPE, new CustomMetaData("data"));
         ClusterState stateWithCustomMetaData = ClusterState.builder(state).metaData(mdBuilder).build();
 
-        final AtomicReference<IllegalStateException> holder = new AtomicReference<>();
+        final CompletableFuture<Throwable> future = new CompletableFuture<>();
         DiscoveryNode node = state.nodes().getLocalNode();
-        zenDiscovery.handleJoinRequest(node, stateWithCustomMetaData, new MembershipAction.JoinCallback() {
-            @Override
-            public void onSuccess() {
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                holder.set((IllegalStateException) e);
-            }
-        });
 
-        assertThat(holder.get(), notNullValue());
-        assertThat(holder.get().getMessage(), equalTo("failure when sending a validation request to node"));
+        transportService.sendRequest(node, PublicationTransportHandler.PUBLISH_STATE_ACTION_NAME,
```

Comment:
this is not the situation where we handle a join anymore but cluster state publishing?
The goal of this test was to make sure that join validation ensures that the node can understand the cluster state before being added to the cluster by the master.

Reply:
Sorry, actually I did not get what the test is supposed to do.
Now it's clear that it should check failing `JoinValidation`.
I still think that `ZenDiscoveryIT` is a better place for this test than `CoordinatorTests` because there is no easy way to force the master node to send state to joining node, which deserialization will fail. However, to have it in `ZenDiscoveryIT` I had to extract one method in `Coordinator` and make it public for tests (similarly we were calling `zenDiscovery.handleJoinRequest` previously).
https://github.com/elastic/elasticsearch/pull/37465/commits/ae39525397fdd781973247789ed9190d9e3cfabc